### PR TITLE
Implement ZCash account creation & balance resolution

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/32305

Adds some functionality related to ZCash:
1) ZCash account creation flow in the webui
2) ZCash balance resolution using lightwalletd grpc [api](https://github.com/zcash/lightwalletd/blob/8003d7fb6e3c47fb0561121a0ee47d080c8c46b9/walletrpc/service.proto#L156) 

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/brave_wallet/browser/zcash/zcash_rpc.cc [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+  replacements.SetPathStr(path);
+
+  return base_url.ReplaceComponents(replacements);
+}
+
+// Prefixes provided serialized protobuf with compression byte and 4 bytes of
+// message size. See
+// https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
+std::string GetPrefixedProtobuf(const std::string& serialized_proto) {
+  char compression = 0;  // 0 means no compression
```
[the remark this answers is not in the document]
Comment (fmarier) [reply components/brave_wallet/browser/zcash/zcash_rpc.cc:71]: False positive: https://github.com/brave/security-action/issues/382

File: components/brave_wallet/browser/keyring_service.cc [hunk 2]
```diff
@@ -2520,6 +2554,56 @@ void KeyringService::HasPendingUnlockRequest(
...
+  auto* zcash_keyring = GetZCashKeyringById(account_id.keyring_id);
+  if (!zcash_keyring) {
+    return absl::nullopt;
+  }
+
+  // TODO(cypt4): store used addresses indexes in prefs.
+
+  // TODO(cypt4): temporarily just return first 5 recieve and 5 change
+  // addresses.
+  std::vector<std::pair<std::string, mojom::ZCashKeyIdPtr>> addresses;
```
Comment (kdenhartog) [inline components/brave_wallet/browser/keyring_service.cc:2570]: I don't think we should assume that the first will always be useful here. Should we check that the addresses are unused via the RPC instead? E.g. someone recovers an already used seed phrase to our wallet then we wouldn't want to reuse these addresses is what I'm thinking.

File: browser/browser_context_keyed_service_factories.cc [hunk 3]
```diff
@@ -150,6 +152,10 @@ void EnsureBrowserContextKeyedServiceFactoriesBuilt() {
   brave_wallet::TxServiceFactory::GetInstance();
   brave_wallet::BraveWalletServiceFactory::GetInstance();
 
+  if (brave_wallet::IsZCashEnabled()) {
+    brave_wallet::ZCashWalletServiceFactory::GetInstance();
+  }
```
Comment (supermassive) [inline browser/browser_context_keyed_service_factories.cc:157]: It seems bitcoin one is missing here. Could please fix this for me

File: components/brave_wallet_ui/components/desktop/popup-modals/add-account-modal/create-account-modal.tsx [hunk 4]
```diff
@@ -210,6 +222,19 @@ export const CreateAccountModal = () => {
               </Select>
             </SelectWrapper>
           }
+          {selectedAccountType?.coin === BraveWallet.CoinType.ZEC &&
+            <SelectWrapper>
+              <Select value={bitcoinNetwork} onChange={onChangeZCashNetwork}>
```
Comment (supermassive) [inline components/brave_wallet_ui/components/desktop/popup-modals/add-account-modal/create-account-modal.tsx:227]: ~bitcoinNetwork~

File: components/brave_wallet_ui/constants/types.ts [hunk 5]
```diff
@@ -792,9 +793,9 @@ export interface TransactionProviderErrorRegistry {
 
 export const SupportedCoinTypes = [
   BraveWallet.CoinType.SOL,
-  BraveWallet.CoinType.ETH,
   BraveWallet.CoinType.FIL,
-  BraveWallet.CoinType.BTC
+  BraveWallet.CoinType.BTC,
+  BraveWallet.CoinType.ZEC
 ]
```
Comment (supermassive) [inline components/brave_wallet_ui/constants/types.ts:799]: Please don't remove ETH :)

File: components/brave_wallet/browser/zcash/zcash_rpc.cc [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    const std::vector<std::string>& addresses) {
+  zcash::GetAddressUtxosRequest request;
+  for (const auto& address : addresses) {
+    request.add_addresses(address);
+  }
+  request.set_maxentries(1);
+  request.set_startheight(0);
+  return GetPrefixedProtobuf(request.SerializeAsString());
+}
+
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_rpc.cc:91]: Is it ok from privacy perspective to send account's addresses in a batch address?

File: components/brave_wallet/browser/zcash/zcash_rpc.cc [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+  }
+  uint32_t size = 0;
+
+  uint8_t as_bytes[4];
+  const char* size_start = &(grpc_response_body.c_str()[1]);
+  for (int i = 0; i < 4; i++) {
+    as_bytes[i] = static_cast<uint8_t>(size_start[i]);
+  }
+  base::ReadBigEndian<uint32_t>(as_bytes, &size);
+
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_rpc.cc:130]:
```
base::ReadBigEndian(reinterpret_cast<const uint8_t*>(&(grpc_response_body[1])), &size);
```

File: components/brave_wallet/browser/zcash/zcash_rpc.cc [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+  return url_loader;
+}
+
+// Resolves serialized protobuf from length-prefixed string
+absl::optional<std::string> ResolveSerializedMessage(
+    const std::string& grpc_response_body) {
+  if (grpc_response_body.size() < 5) {
+    return absl::nullopt;
+  }
+  if (grpc_response_body.c_str()[0] != 0) {
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_rpc.cc:118]: ~.c_str()~

File: components/brave_wallet/browser/zcash/zcash_rpc.cc [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+  for (int i = 0; i < 4; i++) {
+    as_bytes[i] = static_cast<uint8_t>(size_start[i]);
+  }
+  base::ReadBigEndian<uint32_t>(as_bytes, &size);
+
+  if (grpc_response_body.size() != size + 5) {
+    return absl::nullopt;
+  }
+
+  return grpc_response_body.substr(5, grpc_response_body.size() - 5);
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_rpc.cc:135]: just `grpc_response_body.substr(5)` is enough to strip prefix

File: components/brave_wallet/browser/zcash/zcash_rpc.cc [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+      5000);
+}
+
+void ZCashRpc::OnGetUtxosResponse(
+    ZCashRpc::GetUtxoListCallback callback,
+    UrlLoadersList::iterator it,
+    const std::unique_ptr<std::string> response_body) {
+  auto current_loader = std::move(*it);
+  url_loaders_list_.erase(it);
+  zcash::GetAddressUtxosResponse response;
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_rpc.cc:177]: Please move close to first use

File: components/brave_wallet/browser/zcash/zcash_wallet_service.cc [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    context->addresses.insert(address.first);
+  }
+
+  for (const auto& address : context->addresses) {
+    zcash_rpc_->GetUtxoList(
+        chain_id, {address},
+        base::BindOnce(&ZCashWalletService::OnGetUtxosForBalance,
+                       weak_ptr_factory_.GetWeakPtr(), context, address));
+  }
+}
```
Comment (supermassive) [inline components/brave_wallet/browser/zcash/zcash_wallet_service.cc:87]: So do we need batch support in `ZCashRpc::GetUtxoList` ?

File: components/brave_wallet/browser/zcash/zcash_rpc.cc (hunk 6 above)
Comment (kdenhartog) [reply components/brave_wallet/browser/zcash/zcash_rpc.cc:91]: Generally no, although if these are one time use addresses it's less concerning but if we can avoid it we should.

File: components/brave_wallet/browser/keyring_service.cc (hunk 2 above)
Comment (cypt4) [reply components/brave_wallet/browser/keyring_service.cc:2570]: Addresses discovery will be implemented in the future

File: components/resources/wallet_strings.grdp [hunk 12]
```diff
@@ -616,6 +617,7 @@
   <message name="IDS_BRAVE_WALLET_SOL_ACCOUNT_DESCRIPTION" desc="Solana account description">Solana + SVM Chains</message>
   <message name="IDS_BRAVE_WALLET_FIL_ACCOUNT_DESCRIPTION" desc="Filecoin account description">Filecoin</message>
   <message name="IDS_BRAVE_WALLET_BTC_ACCOUNT_DESCRIPTION" desc="Bitcoin account description">Bitcoin</message>
+  <message name="IDS_BRAVE_WALLET_ZEC_ACCOUNT_DESCRIPTION" desc="ZCash account description">ZCash</message>
```
Comment (petemill) [inline components/resources/wallet_strings.grdp:620]: Is ZCash meant to be translated? If not, this xml message should have `translateable="false"`

**Review by petemill — APPROVED** [review]
I only looked at the 2 new strings

File: components/brave_wallet/browser/zcash/zcash_rpc.cc (hunk 6 above)
Comment (cypt4) [reply components/brave_wallet/browser/zcash/zcash_rpc.cc]: Now it takes only 1 address

**Review by Douglashdaniel — APPROVED** [review]
++ Wallet Front-end

**Review by supermassive — APPROVED** [review]

**Comment by nuo-xu** [comment]
can you update `brave_wallet_factory_wrappers` for iOS to have a way to get an instance of `zcash_wallet_service`? i think we need this service to fetch balance. and do we need `zcash_rpc`?

**Review by kdenhartog — APPROVED** [review]
Lgtm

**Comment by cypt4** [comment]
> can you update `brave_wallet_factory_wrappers` for iOS to have a way to get an instance of `zcash_wallet_service`? i think we need this service to fetch balance. and do we need `zcash_rpc`?

You don't need to access ZCashRpc directly

**Review by nuo-xu — APPROVED** [review]

**Comment by nuo-xu** [comment]
iOS ++

**Pull request merged**